File: Programmers/Python/Level 2/뒤에 있는 큰 수 찾기/solution.py
# https://school.programmers.co.kr/learn/courses/30/lessons/154539

def solution(numbers):
    answer = [-1] * len(numbers)
    # 아직 뒤에 큰수가 없는 숫자들의 인덱스
    stack = []
    for idx, num in enumerate(numbers):
        # 현재 인덱스의 숫자가 큰수가 될수있는 인덱스 찾기
        while stack and numbers[stack[-1]] < num:
            answer[stack.pop()] = num
        stack.append(idx)

    return answer

File: Programmers/Javascript/Level 2/뒤에 있는 큰 수 찾기/solution.js
// https://school.programmers.co.kr/learn/courses/30/lessons/154539?language=javascript

function solution(numbers) {
  const stack = [];
  const result = Array(numbers.length).fill(-1);

  numbers.forEach((num, idx) => {
    while (stack.length > 0) {
      const [prevNum, prevIdx] = stack.pop();
      // 현재 숫자가 가장 가까운 큰수가 되는 숫자들 탐색
      if (num > prevNum) {
        result[prevIdx] = num;
      } else {
        stack.push([prevNum, prevIdx]);
        break;
      }
    }
    stack.push([num, idx]);
  });
  return result;
}
